File: Top-Interview-Questions/LinkedList/206-reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 * Runtime: 68 ms, faster than 33.79% of JavaScript online submissions for Reverse Linked List.
 * Memory Usage: 36.8 MB, less than 17.56% of JavaScript online submissions for Reverse Linked List.
 * TC: O(N + N)
 * SC: O(N)
 */
var reverseList = function(head) {
    if (!head) {
        return head;
    }

    const stack = [];
    while (head) {
        stack.push(head);
        head = head.next;
    }

    let reversedHead = stack.pop();
    const reversedHeadCopy = reversedHead;
    while (stack.length !== 0) {
        const current = stack.pop();
        reversedHead.next = current;
        reversedHead = reversedHead.next;
    }

    reversedHead.next = null;

    return reversedHeadCopy;
};

/*
    TC: O(N)
    SC: O(1)
 */
var reverseList = function(head) {
    if (!head) {
        return head;
    }

    let prev = null;
    while (head) {
        const next = head.next;
        
        head.next = prev;
        prev = head;
        head = next;
    }
    
    return prev;
};

/*
    TC: O(N)
    SC: O(N)
 */
var reverseList = function(head) {
    if (!head) {
        return head;
    }
    
    return reverseListHelper(head);
};

function reverseListHelper (node) {
    if (!node.next) {
        return node;
    }
    
    const head = reverseListHelper(node.next);
    node.next.next = node;
    node.next = null;
    
    return head;
}

File: Top-Interview-Questions/LinkedList/206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # Sol 1. Using Stack
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        stack = [] # SC: O(N)
        while head:
            stack.append(head.val)
            head=head.next
            
        #print(stack)
        ret_ll = rev_ll = ListNode()
        #print(ret_ll)
        #print(rev_ll)
        while stack:
        #    print(rev_ll)
            rev_ll.next = ListNode(stack.pop())
            rev_ll = rev_ll.next
        #print(rev_ll)
        #print(ret_ll)
        return ret_ll.next
    # TC: O(n) + O(n) -> O(n), where n is the number of nodes
    # SC: O(n), stack
    # Runtime: 40 ms, faster than 37.80%
    # Memory Usage: 16.1 MB, less than 28.19%

    # Sol 2. Iteration
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        # Change the current node's next pointer to its previous element
        # Singly linked list, we need to store previous value beforehand(prev)
        # Need another pointer(temp) to store the next value
        prev = None
        curr = head
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
            
        return prev
    # TC: O(n)
    # SC: O(1)

    # Sol 3. Recursion
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        rev = self.reverseList(head.next)
        head.next.next = head # place next next node to current node to avoid null
        head.next = None # by doing this the null element will stay at the end of the linked list
        return rev
    # TC: O(n)
    # SC: O(n), recursive function stacks as the amount of elements in the linked list
    
